### ISSUE TITLE
require user consent before hCaptcha inclusion

### ISSUE THREAD

**Issue by benaryorg** [mastodon/mastodon#25025]
### Steps to reproduce the problem

1. a Mastodon instance has the hCaptcha feature turned on
2. I am a user registering on said instance
3. hCaptcha is loaded without consent

### Expected behaviour

user consent should be confirmed

### Actual behaviour

no consent is provided

### Detailed description

If the hCaptcha feature in its present form is to be retained in Mastodon, even temporarily until there is some progress on #25023 there should be some sort of consent to using hCaptcha, not from admins but users.
The registration process consists of a multi-step form already which would make inclusion of a text on the preceding page rather easy I think.
There is of course the privacy policy that should include hCaptcha in such a case, but this feels one step further away than it should, at least from my perspective of a potential user registering on an "Open Platform" (i.e. one associated with FOSS and federation).
Having another checkbox right beneath the checkbox for the privacy policy that requires the user to consent to the use of hCaptcha would be a suitable measure I think, assuming that no hCaptcha connection/code is active at that time.
If this could be added in a way that it is only visible/required with said feature turned on, that would at least be a stop-gap measure that requires explicit user consent before sharing potential PII with a third party.
[The original response in the corresponding Pull Request for the feature](https://github.com/mastodon/mastodon/pull/25019#issuecomment-1551272335) as well as the other mentioned issue provide more context on this.

It would have the added benefit that users with e.g. uMatrix or other tools that block third party content would be informed about the inclusion earlier so they could, in theory, take measures to temporarily allow these third parties (though practically it will require a reload on the next page when the specific domains are known).

### Specifications

Mastodon main branch with commit 67cfbb0 included.
This issue in particular is a front-end issue (the user consent prompt). Other issues with hCaptcha are out of scope for this issue.
All browsers, all operating systems.

**Comment by offbyone** [comment]
I think that the user consent would be expressed by choosing to sign up on an instance that does not have the feature enabled.

**Comment by GiantCrocodile** [comment]
I agree with @stnfrd, I think someone don't understand how GDPR works regarding consent. If you load external stuff before an user agrees on it, you are doing it in 99.999 % of cases simply wrong. It's as simple as that.

**Comment by offbyone** [comment]
I think your understanding of how the internet works and how GDPR works is flawed enough that I'm not sure that it's productive to continue this conversation.

**Comment by GiantCrocodile** [comment]
Feel free to correct me by providing a legal opinion from a recognized expert on this domain or by providing information about a legal opinion of controlling authorities. I'm very interested in this as I would be quite surprised if I'm mistaken on this as I deal with this kind of topics on a daily basis and being connected with several lawyers who deal with this domain.

**Comment by nemobis** [comment]
As described above, there's little doubt that the current usage of hCaptcha is illegal on instances, like mastodon.social, which don't even mention a transfer of personal data to the USA in their [privacy policy](https://mastodon.social/privacy-policy). Better lawyering and paperwork is good as stopgap solution to protect one's ass, however the only real solution is to stop such transfers with a proper replacement: https://github.com/mastodon/mastodon/issues/25023 .
